File: android-agent/src/main/java/io/opentelemetry/android/internal/services/network/detector/PostApi28NetworkDetector.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.internal.services.network.detector;

import static io.opentelemetry.android.internal.services.network.CurrentNetworkProvider.NO_NETWORK;
import static io.opentelemetry.android.internal.services.network.CurrentNetworkProvider.UNKNOWN_NETWORK;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.pm.PackageManager;
import android.net.ConnectivityManager;
import android.net.NetworkCapabilities;
import android.os.Build;
import android.telephony.TelephonyManager;
import androidx.annotation.RequiresApi;
import androidx.core.app.ActivityCompat;
import io.opentelemetry.android.internal.services.network.CarrierFinder;
import io.opentelemetry.android.internal.services.network.data.Carrier;
import io.opentelemetry.android.internal.services.network.data.CurrentNetwork;
import io.opentelemetry.android.internal.services.network.data.NetworkState;

/**
 * This class is internal and not for public use. Its APIs are unstable and can change at any time.
 */
@RequiresApi(api = Build.VERSION_CODES.P)
class PostApi28NetworkDetector implements NetworkDetector {
    private final ConnectivityManager connectivityManager;
    private final TelephonyManager telephonyManager;
    private final CarrierFinder carrierFinder;
    private final Context context;

    PostApi28NetworkDetector(
            ConnectivityManager connectivityManager,
            TelephonyManager telephonyManager,
            CarrierFinder carrierFinder,
            Context context) {
        this.connectivityManager = connectivityManager;
        this.telephonyManager = telephonyManager;
        this.carrierFinder = carrierFinder;
        this.context = context;
    }

    @SuppressLint("MissingPermission")
    @Override
    public CurrentNetwork detectCurrentNetwork() {
        NetworkCapabilities capabilities =
                connectivityManager.getNetworkCapabilities(connectivityManager.getActiveNetwork());
        if (capabilities == null) {
            return NO_NETWORK;
        }
        String subType = null;
        Carrier carrier = carrierFinder.get();
        if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)) {
            // If the app has the permission, use it to get a subtype.
            if (canReadPhoneState()) {
                subType = getDataNetworkTypeName(telephonyManager.getDataNetworkType());
            }
            return CurrentNetwork.builder(NetworkState.TRANSPORT_CELLULAR)
                    .carrier(carrier)
                    .subType(subType)
                    .build();
        } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) {
            return CurrentNetwork.builder(NetworkState.TRANSPORT_WIFI).carrier(carrier).build();
        } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN)) {
            return CurrentNetwork.builder(NetworkState.TRANSPORT_VPN).carrier(carrier).build();
        }
        // there is an active network, but it doesn't fall into the neat buckets above
        return UNKNOWN_NETWORK;
    }

    // visible for testing
    boolean canReadPhoneState() {
        return ActivityCompat.checkSelfPermission(context, Manifest.permission.READ_PHONE_STATE)
                == PackageManager.PERMISSION_GRANTED;
    }

    private String getDataNetworkTypeName(int dataNetworkType) {
        switch (dataNetworkType) {
            case TelephonyManager.NETWORK_TYPE_1xRTT:
                return "1xRTT";
            case TelephonyManager.NETWORK_TYPE_CDMA:
                return "CDMA";
            case TelephonyManager.NETWORK_TYPE_EDGE:
                return "EDGE";
            case TelephonyManager.NETWORK_TYPE_EHRPD:
                return "EHRPD";
            case TelephonyManager.NETWORK_TYPE_EVDO_0:
                return "EVDO_0";
            case TelephonyManager.NETWORK_TYPE_EVDO_A:
                return "EVDO_A";
            case TelephonyManager.NETWORK_TYPE_EVDO_B:
                return "EVDO_B";
            case TelephonyManager.NETWORK_TYPE_GPRS:
                return "GPRS";
            case TelephonyManager.NETWORK_TYPE_GSM:
                return "GSM";
            case TelephonyManager.NETWORK_TYPE_HSDPA:
                return "HSDPA";
            case TelephonyManager.NETWORK_TYPE_HSPA:
                return "HSPA";
            case TelephonyManager.NETWORK_TYPE_HSPAP:
                return "HSPAP";
            case TelephonyManager.NETWORK_TYPE_HSUPA:
                return "HSUPA";
            case TelephonyManager.NETWORK_TYPE_IDEN:
                return "IDEN";
            case TelephonyManager.NETWORK_TYPE_IWLAN:
                return "IWLAN";
            case TelephonyManager.NETWORK_TYPE_LTE:
                return "LTE";
            case TelephonyManager.NETWORK_TYPE_NR:
                return "NR";
            case TelephonyManager.NETWORK_TYPE_TD_SCDMA:
                return "SCDMA";
            case TelephonyManager.NETWORK_TYPE_UMTS:
                return "UMTS";
            case TelephonyManager.NETWORK_TYPE_UNKNOWN:
                return "UNKNOWN";
        }
        return "UNKNOWN";
    }
}

File: services/src/main/java/io/opentelemetry/android/internal/services/network/detector/PostApi28NetworkDetector.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.internal.services.network.detector

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.PackageManager
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import android.telephony.TelephonyManager
import android.telephony.TelephonyManager.*
import androidx.annotation.RequiresApi
import androidx.core.app.ActivityCompat
import io.opentelemetry.android.common.internal.features.networkattributes.data.Carrier
import io.opentelemetry.android.common.internal.features.networkattributes.data.CurrentNetwork
import io.opentelemetry.android.common.internal.features.networkattributes.data.NetworkState
import io.opentelemetry.android.common.internal.features.networkattributes.data.NetworkState.TRANSPORT_CELLULAR
import io.opentelemetry.android.common.internal.features.networkattributes.data.NetworkState.TRANSPORT_VPN
import io.opentelemetry.android.common.internal.features.networkattributes.data.NetworkState.TRANSPORT_WIFI
import io.opentelemetry.android.internal.services.network.CarrierFinder
import io.opentelemetry.android.internal.services.network.CurrentNetworkProvider
import io.opentelemetry.android.internal.services.network.CurrentNetworkProvider.UNKNOWN_NETWORK

/**
 * This class is internal and not for public use. Its APIs are unstable and can change at any time.
 */
@RequiresApi(api = Build.VERSION_CODES.P)
//TODO: Refactor tests and remove open
internal open class PostApi28NetworkDetector(
    private val connectivityManager: ConnectivityManager,
    private val telephonyManager: TelephonyManager,
    private val carrierFinder: CarrierFinder,
    private val context: Context
) : NetworkDetector {
    
    @SuppressLint("MissingPermission")
    override fun detectCurrentNetwork(): CurrentNetwork {
        val capabilities =
            connectivityManager.getNetworkCapabilities(connectivityManager.activeNetwork)
                ?: return CurrentNetworkProvider.NO_NETWORK
        val carrier = carrierFinder.get()

        fun hasTransport(capability: Int): Boolean = capabilities.hasCapability(capability)

        val network = buildFromCarrier(carrier)
        return when {
            hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> withCellDataType(carrier)
            hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> network(TRANSPORT_WIFI)
            hasTransport(NetworkCapabilities.TRANSPORT_VPN) -> network(TRANSPORT_VPN)
            else -> UNKNOWN_NETWORK
        }
    }

    private fun buildFromCarrier(carrier: Carrier): (NetworkState) -> CurrentNetwork =
        { CurrentNetwork.builder(it).carrier(carrier).build() }


    private fun withCellDataType(
        carrier: Carrier
    ): CurrentNetwork {
        val builder = CurrentNetwork.builder(TRANSPORT_CELLULAR)
            .carrier(carrier)
        if (canReadPhoneState()) {
            val dataNetworkType = getDataNetworkTypeName(telephonyManager.dataNetworkType)
            return builder.subType(dataNetworkType).build()
        }
        return builder.build();
    }

    // visible for testing
    open fun canReadPhoneState(): Boolean {
        return (ActivityCompat.checkSelfPermission(context, Manifest.permission.READ_PHONE_STATE)
                == PackageManager.PERMISSION_GRANTED)
    }

    private fun getDataNetworkTypeName(dataNetworkType: Int): String {
        return when (dataNetworkType) {
            NETWORK_TYPE_1xRTT -> "1xRTT"
            NETWORK_TYPE_CDMA ->  "CDMA"
            NETWORK_TYPE_EDGE ->  "EDGE"
            NETWORK_TYPE_EHRPD ->  "EHRPD"
            NETWORK_TYPE_EVDO_0 ->  "EVDO_0"
            NETWORK_TYPE_EVDO_A ->  "EVDO_A"
            NETWORK_TYPE_EVDO_B ->  "EVDO_B"
            NETWORK_TYPE_GPRS ->  "GPRS"
            NETWORK_TYPE_GSM ->  "GSM"
            NETWORK_TYPE_HSDPA ->  "HSDPA"
            NETWORK_TYPE_HSPA ->  "HSPA"
            NETWORK_TYPE_HSPAP ->  "HSPAP"
            NETWORK_TYPE_HSUPA ->  "HSUPA"
            NETWORK_TYPE_IDEN ->  "IDEN"
            NETWORK_TYPE_IWLAN ->  "IWLAN"
            NETWORK_TYPE_LTE ->  "LTE"
            NETWORK_TYPE_NR ->  "NR"
            NETWORK_TYPE_TD_SCDMA ->  "SCDMA"
            NETWORK_TYPE_UMTS ->  "UMTS"
            NETWORK_TYPE_UNKNOWN ->  "UNKNOWN"
            else -> "UNKNOWN"
        }
    }

    private fun network(state: NetworkState): CurrentNetwork = CurrentNetwork.builder(state).build()
}
